packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,306 @@
+    env->debug(DETAILED_STATUS, "	graph model built");
+    
+    //Get vertex GIDs, in a locally indexed array
+    ArrayView<const gno_t> vtxIDs;
+    ArrayView<StridedData<lno_t, scalar_t> > vwgts;
+    size_t nVtx = model->getVertexList(vtxIDs, vwgts);
+    //at this point, weights are not used.
+ 
+    //Get the edge information
+    ArrayView<const gno_t> adjs; 
+    ArrayView<const offset_t> offsets;
+    ArrayView<StridedData<lno_t, scalar_t> > ewgts;
+    size_t nEdge = model->getEdgeList(adjs, offsets, ewgts);
+    //edge weights are not used either.
+  
+    //copy data over to regular arrays
+    gno_t* out_edges = NULL;
+    typename map_t::local_ordinal_type* out_offsets = NULL;
+    gno_t* global_ids = NULL;
+    TPL_Traits<gno_t, const gno_t>::ASSIGN_ARRAY(&out_edges, adjs);
+    TPL_Traits<typename map_t::local_ordinal_type, const offset_t>::ASSIGN_ARRAY(&out_offsets, offsets);
+    TPL_Traits<gno_t, const gno_t>::ASSIGN_ARRAY(&global_ids, vtxIDs);
+  
+    //get the rank of the current process
+    int me = problemComm->getRank();
+  
+    //create the Tpetra map for the global indices
+    Tpetra::global_size_t dummy = Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid();
+    RCP<const map_t > map = rcp(new map_t(dummy, vtxIDs, 0, problemComm)); 
+  
+    //sentinel value to detect unsuccessful searches
+    lno_t fail = Teuchos::OrdinalTraits<lno_t>::invalid();
+
+    //map to make sure ghosted vertices aren't duplicated in the Tpetra Maps
+    std::unordered_map<gno_t,lno_t> ghost_map;
+    //The new local IDs for each ghosted vertex
+    std::vector<lno_t> ghostGIDs;
+    int nGhosts = 0;
+    //count how many ghosts are in the edge list, to create the mapWithCopies.
+    for(size_t i = 0; i < nEdge; i++){
+      if(map->getLocalElement(adjs[i]) == fail){
+        if(ghost_map.count(adjs[i]) == 0){
+          ghost_map[adjs[i]] = nVtx + nGhosts;
+          ghostGIDs.push_back(adjs[i]);
+          nGhosts++;
+        }
+      }
+    }
+  
+    //build nVtx + nGhosts size array of grounding flags
+    bool* grounding = new bool[nVtx+nGhosts];
+    for(size_t i = 0; i < nVtx+nGhosts; i++){
+      if(i < nVtx){
+        grounding[i] = basalFriction[i];
+      }
+      else grounding[i] = false;
+    }
+
+  
+    //use the count of ghosts + owned to make mapWithCopies, need to create a Teuchos array first.
+    Teuchos::Array<gno_t> gids(nVtx+nGhosts);
+    for(size_t i = 0; i < nVtx; i++){
+      gids[i] = vtxIDs[i];
+    }
+    for(size_t i = nVtx; i < nVtx+nGhosts; i++){
+      gids[i] = ghostGIDs[i-nVtx];
+    }
+  
+    //create the Tpetra map with copies
+    Tpetra::global_size_t dummy_2 = Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid();
+    RCP<const map_t > mapWithCopies = rcp(new map_t(dummy_2, gids, 0, problemComm));
+ 
+    //communicate boundary edges back to owning processors
+    gno_t num_boundary_edges = boundary_edges.size()/2;
+    std::vector<int> owners_vec;
```

Comment:
Why are the copies owners_vec and boundary_vec needed?
Why not use boundary_edges directly instead of making boundary_vec and boundary?
Why not declare Teuchos::Array owners(boundary_edges.size(), 0)?

Reply:
I wasn't aware of a better way to do this, and there are likely many places that I have clumsy usage of Teuchos datastructures. I will clean this up as you suggest.